packages/server/src/services/settingsService.ts

```
@@ -0,0 +1,74 @@
+import fs from 'fs';
+import path from 'path';
+
+import { AttestationFormat } from '../helpers/decodeAttestationObject';
+import convertCertBufferToPEM from '../helpers/convertCertBufferToPEM';
+
+class SettingsService {
+  // Certificates are stored as PEM-formatted strings
+  private pemCertificates: Map<AttestationFormat, string>;
+
+  constructor() {
+    this.pemCertificates = new Map();
+  }
+
+  /**
+   * Allow setting custom root certificates for attestation formats that use them
+   *
+   * The certificate can be specified as a raw `Buffer`, or as a PEM-formatted string. If a
+   * `Buffer` is passed in it will be converted to PEM format.
+   */
+  setRootCertificate(opts: {
+    attestationFormat: AttestationFormat;
+    certificate: Buffer | string;
+  }): void {
+    const { attestationFormat } = opts;
+    let { certificate: newCertificate } = opts;
+
+    if (newCertificate instanceof Buffer) {
+      newCertificate = convertCertBufferToPEM(newCertificate);
+    }
+
+    this.pemCertificates.set(attestationFormat, newCertificate);
+  }
+
+  /**
+   * Get any registered root certificates for the specified attestation format
+   */
+  getRootCertificate(opts: { attestationFormat: AttestationFormat }): string {
+    const { attestationFormat } = opts;
+    return this.pemCertificates.get(attestationFormat) ?? '';
```

Why return `''` as a sentinel value?

I'm not sure what happens if an empty string is pushed onto the certificate chain in the verify functions, but my guess is that it will fail with an error that won't clearly point to the problem being that a root certificate wasn't registered.

My initial thought was that this should throw an error if `!this.pemCertificates.has(attestationFormat)`, but I'm not sure if there is some valid case for requesting a certificate which has not been registered. If there is, then explicitly registering `null` and checking for `null` in the verify functions might be more robust.

---

I went this route for cleaner typing, if the method is guaranteed to return a string then that's one less type guard I'd need to write 😅 

My gut tells me it's better to just return whatever `.get()` returns, which would make this `string | undefined`. Then the verification method can be updated to check the value of `rootCertificate` and impose its restrictions on what should have been returned.

> ...my guess is that it will fail with an error that won't clearly point to the problem being that a root certificate wasn't registered.

The cert path validation code would likely error out that the value provided wasn't a valid certificate structure. I kinda think if you're using this feature you'll be smart enough to know how to properly specify a PEM cert, since this is a Power User feature.

---

Since tweaking SettingsService to support setting multiple possible root certs this method has been updated to return an empty array instead. It makes sense to me to do so, since the logic later on that uses these values checks if `rootCertificates.length === 0` and branches appropriately.